### PR TITLE
feat: implement account mute/unmute API with persistent storage

### PR DESCRIPTION
**Pull request by llun**
## Summary

- Add `mutes` table migration with `notifications` (bool) and `endsAt` (Unix ms bigint) columns; no AP federation, local-only
- Add `MuteDatabase` interface + `MuteSQLDatabaseMixin` (idempotent create/update, delete, isMuting, getMuteRelations); wire into `Database` type and SQL index
- Replace hardcoded `muting: false` in relationship service with real `getMute` lookup; implement `POST /api/v1/accounts/:id/mute` (body: `notifications`, `duration`) and `/unmute` route handlers

## Test Plan

- [x] `yarn build` passes (no TypeScript errors)
- [x] `yarn lint` passes (0 errors)
- [x] All 2884 tests pass (`yarn test`)
- [x] New `lib/database/sql/mute.test.ts` covers create, re-mute update, delete, isMuting, getMuteRelations directional filtering, empty-input edge cases
- [x] `lib/services/accounts/relationship.test.ts` extended with muting=true/false and muting_notifications=true/false test cases
- [ ] Manual: `POST /api/v1/accounts/:id/mute` → `muting: true, muting_notifications: true`
- [ ] Manual: `POST /api/v1/accounts/:id/mute` with `{"notifications": false}` → `muting_notifications: false`
- [ ] Manual: `GET /api/v1/accounts/relationships?id[]=:id` reflects mute state
- [ ] Manual: `POST /api/v1/accounts/:id/unmute` → `muting: false`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by llun** [comment]
## Code Review Response

Addressed in commit 1cf82d00:

### Fixed ✅

**Critical #3 — Mute orphans on actor deletion**
Added `mutes` table cleanup in `deleteActorData` transaction alongside the existing blocks cleanup (`lib/database/sql/actor.ts`).

**Important #4 — No route-level tests**
Added `app/api/v1/accounts/[id]/mute/route.test.ts` and `…/unmute/route.test.ts` covering: default params, `notifications: false`, `duration` → `endsAt` computation, negative/fractional duration, self-mute no-op, and empty-body edge case.

**Important #5 — Duration validation**
Added `Math.floor()` to reject fractional seconds and `> 0` guard to treat negative values as no-expiry (`app/api/v1/accounts/[id]/mute/route.ts`).

**Minor #7 — `endsAt` truthiness check**
Changed to explicit `!== null && !== undefined` check in `fixMuteDataDate`.

---

### Not addressed — with reasoning

**Critical #1 — Muting has no effect on timeline/notification delivery**
The plan explicitly scopes only the API persistence layer (DB, types, routes, relationship service). Adding timeline/notification filtering would require updating `blockFilter.ts`, `shouldNotify.ts`, and potentially `createNote.ts` — a separate feature slice. Tracking as follow-up work.

**Critical #2 — `endsAt` not enforced in queries**
The plan explicitly says: *"storage only; auto-expiry is out of scope"*. The `endsAt` column is stored for future use. Adding a live `WHERE endsAt IS NULL OR endsAt > now()` filter is the correct approach when auto-expiry is in scope, but adding it now would conflict with the documented design decision.

**Important #6 — `this` binding fragility in mixin**
This is a pre-existing pattern throughout the codebase — `block.ts` uses `this.getBlock(...)` in exactly the same way. Fixing this here but not in block would be inconsistent. Filed as a codebase-wide refactor concern.

**Minor #8 — Missing `read:mutes` / `write:mutes` scopes**
Pre-existing scope system limitation, not introduced by this PR. The block/unblock routes use the same `write` catch-all. Out of scope.

**Minor #9 — No actor-existence check before operating**
Mutes are local-only (no federation side-effects). A phantom mute row for an unknown actor ID is harmless and `getRelationship` handles the null actor gracefully. Adding a 404 guard would diverge from Mastodon's documented behaviour of accepting mute calls for any ID.

File: lib/database/sql/mute.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+const fixMuteDataDate = (data: Mute): Mute => ({
+  ...data,
+  notifications: Boolean(data.notifications),
+  endsAt: data.endsAt ? Number(data.endsAt) : null,
+  createdAt: getCompatibleTime(data.createdAt),
+  updatedAt: getCompatibleTime(data.updatedAt)
+})
+
+const MUTE_RELATION_LOOKUP_CHUNK_SIZE = 1000
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts]: Fixed in latest commit: reduced MUTE_RELATION_LOOKUP_CHUNK_SIZE from 1000 to 400. With two WHERE IN clauses per query (actorId + targetActorId), the combined parameter count is 400×2=800, safely under SQLite's 999-parameter limit.

File: lib/database/sql/mute.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    return fixMuteDataDate(existingMute)
+  },
+
+  async getMute({ actorId, targetActorId }: GetMuteParams) {
+    const mute = await database<Mute>('mutes')
+      .where({ actorId, targetActorId })
+      .first()
+    if (!mute) return null
+    return fixMuteDataDate(mute)
+  },
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts:133]: Fixed: getMute now checks if endsAt is in the past and returns null for expired mutes. Added test coverage for this case.

File: lib/database/sql/mute.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+  async isMuting({ actorId, targetActorId }: IsMutingParams) {
+    const mute = await database('mutes')
+      .where({ actorId, targetActorId })
+      .first('id')
+    return Boolean(mute)
+  },
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts:146]: Fixed: isMuting now fetches endsAt and returns false when the mute has expired. Added test coverage.

File: lib/database/sql/mute.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    )
+
+    for (const relations of relationGroups) {
+      for (const relation of relations) {
+        relationsByKey.set(
+          JSON.stringify([relation.actorId, relation.targetActorId]),
+          { ...relation, notifications: Boolean(relation.notifications) }
+        )
+      }
+    }
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts]: Fixed: getMuteRelations now executes chunks sequentially (nested for loops instead of Promise.all/flatMap) and filters expired mutes by checking endsAt per row. Test added.

File: app/api/v1/accounts/[id]/mute/route.ts [hunk 5]
```diff
@@ -29,8 +30,21 @@ export const POST = traceApiRoute(
...
 
-    // Muting not yet implemented - return relationship with muting: false
-    // TODO: Implement muting functionality
+    if (targetActorId !== currentActor.id) {
+      const body = await req.json().catch(() => ({}))
+      const notifications: boolean = body.notifications !== false
+      const durationSeconds =
+        typeof body.duration === 'number' ? body.duration : 0
+      const endsAt =
+        durationSeconds > 0 ? Date.now() + durationSeconds * 1000 : null
```
[the remark this answers is not in the document]
Comment (llun) [reply app/api/v1/accounts/[id]/mute/route.ts]: Fixed: replaced manual inline checks with a Zod schema (MuteBodySchema) that validates notifications as boolean and duration as a finite non-negative number ≤ 3153600000 with Math.floor transform. Uses safeParse so invalid/missing body gracefully falls back to defaults.

File: app/api/v1/accounts/[id]/unmute/route.ts [hunk 6]
```diff
@@ -29,8 +30,9 @@ export const POST = traceApiRoute(
 
     const targetActorId = idToUrl(encodedAccountId)
 
-    // Unmuting not yet implemented - return relationship with muting: false
-    // TODO: Implement muting functionality
+    if (targetActorId !== currentActor.id) {
+      await applyUnmute({ database, actorId: currentActor.id, targetActorId })
+    }
```
[the remark this answers is not in the document]
Comment (llun) [reply app/api/v1/accounts/[id]/unmute/route.ts:57]: Fixed: unmute route now mirrors the unblock pattern — if no mute record exists, we check whether the target actor exists and return 404 if it doesn't. Unmuting a known actor who wasn't muted remains a no-op.

File: lib/database/sql/mute.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  IsMutingParams,
+  MuteDatabase,
+  MuteRelation
+} from '@/lib/types/database/operations'
+import { Mute } from '@/lib/types/domain/mute'
+
+const fixMuteDataDate = (data: Mute): Mute => ({
+  ...data,
+  notifications: Boolean(data.notifications),
+  endsAt: data.endsAt ? Number(data.endsAt) : null,
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts]: Fixed before this review: fixMuteDataDate already uses an explicit null/undefined check (`data.endsAt !== null && data.endsAt !== undefined ? Number(data.endsAt) : null`) so a valid 0 round-trips correctly.

File: lib/database/sql/mute.ts [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+    await database('mutes').where('id', existingMute.id).delete()
+    return fixMuteDataDate(existingMute)
+  },
+
+  async getMute({ actorId, targetActorId }: GetMuteParams) {
+    const mute = await database<Mute>('mutes')
+      .where({ actorId, targetActorId })
+      .first()
+    if (!mute) return null
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts:129]: Fixed: getMute now returns null when endsAt is in the past, so callers see expired mutes as no mute. Test added.

File: lib/database/sql/mute.ts [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    return fixMuteDataDate(mute)
+  },
+
+  async isMuting({ actorId, targetActorId }: IsMutingParams) {
+    const mute = await database('mutes')
+      .where({ actorId, targetActorId })
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts:137]: Fixed: isMuting fetches endsAt alongside id and returns false when the mute has expired. Test added.

File: lib/database/sql/mute.ts [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+      MUTE_RELATION_LOOKUP_CHUNK_SIZE
+    )
+
+    const relationGroups = await Promise.all(
+      actorIdChunks.flatMap((actorIdChunk) =>
+        targetActorIdChunks.map((targetActorIdChunk) =>
+          database<MuteRelation>('mutes')
+            .select('actorId', 'targetActorId', 'notifications')
+            .whereIn('actorId', actorIdChunk)
+            .whereIn('targetActorId', targetActorIdChunk)
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts]: Fixed: getMuteRelations now includes endsAt in the SELECT, filters out any relation where endsAt is in the past, and runs chunks sequentially. Test added.

File: app/api/v1/accounts/[id]/mute/route.ts [hunk 11]
```diff
@@ -29,8 +30,21 @@ export const POST = traceApiRoute(
 
     const targetActorId = idToUrl(encodedAccountId)
 
-    // Muting not yet implemented - return relationship with muting: false
-    // TODO: Implement muting functionality
+    if (targetActorId !== currentActor.id) {
+      const body = await req.json().catch(() => ({}))
+      const notifications: boolean = body.notifications !== false
```
[the remark this answers is not in the document]
Comment (llun) [reply app/api/v1/accounts/[id]/mute/route.ts:64]: Fixed: mute route now calls database.getActorFromId before applying the mute; returns 404 if the target actor is not found locally. Test added.

File: lib/database/sql/mute.ts [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+  notifications: Boolean(data.notifications),
+  endsAt:
+    data.endsAt !== null && data.endsAt !== undefined
+      ? Number(data.endsAt)
+      : null,
+  createdAt: getCompatibleTime(data.createdAt),
+  updatedAt: getCompatibleTime(data.updatedAt)
+})
+
+const MUTE_RELATION_LOOKUP_CHUNK_SIZE = 1000
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts]: Fixed: chunk size reduced to 400 (same fix as the first review comment above).

File: lib/database/sql/mute.ts [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+    const relationGroups = await Promise.all(
+      actorIdChunks.flatMap((actorIdChunk) =>
+        targetActorIdChunks.map((targetActorIdChunk) =>
+          database<MuteRelation>('mutes')
+            .select('actorId', 'targetActorId', 'notifications')
+            .whereIn('actorId', actorIdChunk)
+            .whereIn('targetActorId', targetActorIdChunk)
+        )
+      )
+    )
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts]: Fixed: switched to sequential nested for loops so chunks execute one at a time, consistent with the guidance to avoid unbounded concurrent DB load.

File: lib/database/sql/mute.ts [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+export const MuteSQLDatabaseMixin = (database: Knex): MuteDatabase => ({
+  async createMute({
+    actorId,
+    targetActorId,
+    notifications,
+    endsAt
+  }: CreateMuteParams) {
+    const currentTime = new Date()
+
+    const existingMute = await this.getMute({ actorId, targetActorId })
```
Comment (llun) [inline lib/database/sql/mute.ts]: **Critical: re-muting a target whose previous mute has expired throws a 500.**

`getMute` (line 117–125) filters out expired rows and returns `null`. That means when an expired mute row for `(actorId, targetActorId)` still exists in the table:

1. Line 52 calls `this.getMute(...)` → returns `null` (row exists but is expired).
2. Code falls through to the INSERT path (line 79–85).
3. INSERT trips the `mutes_actor_target_unique` constraint on `(actorId, targetActorId)` and throws.
4. The catch on line 86 calls `this.getMute(...)` again on line 89 → still `null` for the same reason.
5. `throw error` on line 103 re-throws the unique-constraint error to the caller.

User-visible result: `POST /api/v1/accounts/:id/mute` returns 500 whenever the previous mute on that target has expired but the row hasn't been cleaned up — and there is no cleanup path (the unmute route also early-returns when `getMute` reports `null`, so the row never gets removed).

Suggested fix: in both the pre-check (line 52) and the catch recovery (line 89), look up the raw row directly without the expiry filter — e.g.
```ts
const existingRow = await database<Mute>('mutes')
  .where({ actorId, targetActorId })
  .first()
```
Then, if a row exists (expired or not), UPDATE `notifications`, `endsAt`, and `updatedAt` to revive/refresh it. This both fixes the 500 and gives expired rows a natural cleanup path.

The existing test "updates notifications and endsAt when re-muting the same target" doesn't catch this because both mutes in that test have non-expired `endsAt`. Worth adding a regression test that inserts a row with `endsAt` in the past and then calls `createMute` again.

Comment (llun) [reply lib/database/sql/mute.ts]: Fixed. createMute now uses a raw DB lookup (without the expiry filter) so expired rows are detected and UPDATEd rather than triggering a unique-constraint INSERT error. The unmute route was also simplified: it calls applyUnmute (deleteMute is raw) directly and only returns 404 when nothing was deleted AND the actor is unknown — which also cleans up expired rows when a user explicitly unmutes. Regression test added.

File: lib/database/sql/mute.ts [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+        })
+        return {
+          ...duplicated,
+          notifications,
+          endsAt,
+          updatedAt: currentTime.getTime()
+        }
+      }
+      throw error
+    }
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts]: This was already fixed in commit f4e3882c (round-1 fixes) and confirmed in f17f7352. createMute now uses a raw Knex query (`database<Mute>('mutes').where({ actorId, targetActorId }).first()`) instead of `this.getMute()`, so expired rows are detected and UPDATEd rather than causing a constraint violation. Regression test added at mute.test.ts:174–208.

File: app/api/v1/accounts/[id]/mute/route.ts [hunk 16]
```diff
@@ -29,8 +48,30 @@ export const POST = traceApiRoute(
...
+      if (!targetActor)
+        return apiResponse({
+          req,
+          allowedMethods: CORS_HEADERS,
+          data: ERROR_404,
+          responseStatusCode: 404
+        })
+
+      const rawBody = await req.json().catch(() => ({}))
+      const body = MuteBodySchema.safeParse(rawBody).data ?? {}
```
[the remark this answers is not in the document]
Comment (llun) [reply app/api/v1/accounts/[id]/mute/route.ts:63]: Intentionally lenient here for Mastodon API compatibility.

Both `notifications` and `duration` are optional parameters — a client that calls `POST /api/v1/accounts/:id/mute` without a body is a valid and common pattern (e.g., Mastodon's own Android client). Returning 400 for an empty body would break these clients unnecessarily.

The current approach mirrors Mastodon's reference implementation:
- `req.json().catch(() => ({}))` handles the empty-body case gracefully (no body → use defaults)
- `MuteBodySchema.safeParse(rawBody).data ?? {}` ignores unknown fields and uses defaults when validation fails

This is explicitly covered by the "handles empty body gracefully" test, which documents the intentional contract. The Zod schema already validates and transforms the fields that are present; extra or malformed fields are simply ignored and defaults are used.

If the body contained structurally invalid JSON that is also non-empty (e.g. a binary blob), it would still degrade to defaults rather than 400 — acceptable for this optional-parameter endpoint.

File: app/api/v1/accounts/[id]/mute/route.ts [hunk 17]
```diff
@@ -1,15 +1,34 @@
...
+const MuteBodySchema = z.object({
+  notifications: z.boolean().optional(),
+  duration: z
+    .number()
+    .finite()
+    .min(0)
+    .max(3_153_600_000)
+    .transform(Math.floor)
+    .optional()
+})
```
[the remark this answers is not in the document]
Comment (llun) [reply app/api/v1/accounts/[id]/mute/route.ts:31]: Fixed in f41da6d3. Added `.catch()` on each field individually so that an invalid `duration` (e.g. negative, Infinity) falls back to `0` without discarding the rest of the object, and an invalid `notifications` falls back to `true`. This preserves `notifications: false` even when `duration` is invalid.

Added a regression test: `'preserves valid notifications:false when duration is invalid'` to cover exactly this scenario.

File: lib/database/sql/mute.ts [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+          .whereIn('targetActorId', targetActorIdChunk)
+
+        for (const relation of relations) {
+          const endsAt =
+            relation.endsAt !== null && relation.endsAt !== undefined
+              ? Number(relation.endsAt)
+              : null
+          if (endsAt !== null && endsAt < now) continue
+          relationsByKey.set(
+            JSON.stringify([relation.actorId, relation.targetActorId]),
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts:186]: Not applying this suggestion for two reasons:

1. **`normalizeActorId` does not exist in this codebase** — introducing it would fail TypeScript compilation. There is no such utility anywhere under `lib/`.

2. **This pattern mirrors the established convention in `block.ts`** — the existing `getBlockRelations` implementation at `lib/database/sql/block.ts:274` uses the identical `JSON.stringify([relation.actorId, relation.targetActorId])` key. The mute mixin was intentionally written to match that pattern for consistency.

On correctness: actor IDs in this DB column are stored once from the canonical form (e.g. `https://remote.test/users/alice`) and always fetched back verbatim, so raw string equality is sufficient — no normalization is needed.

On performance: the Map is built per-request from a bounded in-memory result set (typical users have O(10²) mutes at most), so the serialization overhead of a 2-element array stringify is negligible compared to the DB round-trip.

File: lib/database/sql/mute.ts [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+      id: randomUUID(),
+      actorId,
+      actorHost: new URL(actorId).host,
+      targetActorId,
+      targetActorHost: new URL(targetActorId).host,
+      notifications,
+      endsAt,
+      createdAt: currentTime.getTime(),
+      updatedAt: currentTime.getTime()
+    }
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/database/sql/mute.ts:82]: Not applying — this pattern is identical to the established convention in `block.ts` (line 72–74) and `follow.ts` (line 59–61), both of which use `new URL(actorId).host` without a try-catch.

By the time `createMute` is called:
- `actorId` is the authenticated actor's ID, sourced from the session — always a valid URL
- `targetActorId` has already been validated against the database (`getActorFromId` in the route handler returns 404 before `applyMute` is invoked)

Wrapping just `mute.ts` but not `block.ts`/`follow.ts` would introduce inconsistency. If this defensive pattern is desired project-wide, it belongs in a separate refactor of all three mixins together.
